src/h3lib/lib/iterators.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2021 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file Iter_Child.c
+ * @brief Iterator structs and functions for the children of a cell,
+ * or cells at a given resolution.
+ */
+
+#include "iterators.h"
+
+#include "h3Index.h"
+
+/*
+extract the `res` digit (0--7) of the current cell
+ */
+static int _get(Iter_Child* it, int res) {
+    int s = 3 * (15 - res);
+    uint64_t m = 7;
+
+    m <<= s;
+
+    return (it->h & m) >> s;
+}
+
+/*
+increment the digit (0--7) at location `res`
+ */
+static void _inc(Iter_Child* it, int res) {
+    uint64_t val = 1;
+    val <<= 3 * (15 - res);
+    it->h += val;
+}
+
+/*
+Fully nulled-out child iterator for when an iterator is exhausted.
+This helps minimize the chance that a user will depend on the iterator
+internal state after it's exhausted, like the child resolution, for
+example.
+ */
+static Iter_Child _null_iter() {
+    return (Iter_Child){.h = H3_NULL, .parentRes = -1, .fnz = -1};
```

Comment:
The cast operator should not be needed here?

Reply:
Seems like it is needed:

```
/Users/ajfriend/work/h3/src/h3lib/lib/iterators.c:48:12: error: expected expression [clang-diagnostic-error]
    return {.h = H3_NULL, .parentRes = -1, .fnz = -1};
           ^
1 error generated.
Error while processing /Users/ajfriend/work/h3/src/h3lib/lib/iterators.c.
```